File: 04-Linked-List/Odd-Even-Linked-List/solution.py
"""
Problem      : Odd Even Linked List
Platform     : LeetCode
Difficulty   : Medium

Pattern       : Linked List, Two Pointers
Time          : O(n)
Space         : O(1)

Author        : Prakritish Bhattacharya
GitHub        : github.com/Prakritish-bhattacharya
"""
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def oddEvenList(self, head):
        if not head or not head.next:
            return head

        odd = head
        even = head.next
        evenStart = even

        while even and even.next:
            odd.next = odd.next.next
            even.next = even.next.next

            odd = odd.next
            even = even.next

        odd.next = evenStart

        return head


# -------------------------
# Helper Functions
# -------------------------

def createLinkedList(arr):
    if not arr:
        return None

    head = ListNode(arr[0])
    curr = head

    for value in arr[1:]:
        curr.next = ListNode(value)
        curr = curr.next

    return head


def printLinkedList(head):
    while head:
        print(head.val, end=" -> ")
        head = head.next
    print("None")


# -------------------------
# Driver Code
# -------------------------

if __name__ == "__main__":

    arr = [1, 2, 3, 4, 5]

    head = createLinkedList(arr)

    print("Original List:")
    printLinkedList(head)

    solution = Solution()
    head = solution.oddEvenList(head)

    print("\nAfter Rearranging:")
    printLinkedList(head)

File: 04-Linked-List/Odd-Even-Linked-List/solution.js
/*
Problem      : Odd Even Linked List
Platform     : LeetCode
Difficulty   : Medium

Pattern       : Linked List, Two Pointers
Time          : O(n)
Space         : O(1)

Author        : Prakritish Bhattacharya
GitHub        : github.com/Prakritish-bhattacharya
*/

class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    oddEvenList(head) {
        if (!head || !head.next) return head;

        let odd = head;
        let even = head.next;
        const evenStart = even;

        while (even && even.next) {
            odd.next = odd.next.next;
            even.next = even.next.next;

            odd = odd.next;
            even = even.next;
        }

        odd.next = evenStart;

        return head;
    }
}

// -------------------------
// Helper Functions
// -------------------------

function createLinkedList(arr) {
    if (arr.length === 0) return null;

    const head = new ListNode(arr[0]);
    let curr = head;

    for (let i = 1; i < arr.length; i++) {
        curr.next = new ListNode(arr[i]);
        curr = curr.next;
    }

    return head;
}

function printLinkedList(head) {
    let curr = head;

    while (curr) {
        process.stdout.write(curr.val + " -> ");
        curr = curr.next;
    }

    console.log("null");
}

// -------------------------
// Driver Code
// -------------------------

const arr = [1, 2, 3, 4, 5];

let head = createLinkedList(arr);

console.log("Original List:");
printLinkedList(head);

const solution = new Solution();
head = solution.oddEvenList(head);

console.log("\nAfter Rearranging:");
printLinkedList(head);